### PR TITLE
Use the API for inline javascript on email clocking

### PR DESCRIPTION
**Pull request by dgrammatiko**
#### Summary of Changes

Using the API for inline script
#### Testing Instructions

Create a new article and write an email address in it. save it and make a menu so you can see it in the front end.
Make sure the plugin email cloacking is enabled

Apply patch
Visit the page you should still see the email address
Disable the javascript in your browser and refresh the page. You should see the next message:
This email address is being protected from spambots. You need JavaScript enabled to view it.

Useful link: https://developer.mozilla.org/en-US/docs/Web/Events/readystatechange

### CODE REVIEW

File: libraries/cms/html/email.php [hunk 1]
```diff
@@ -75,32 +67,57 @@ public static function cloak($mail, $mailto = true, $text = '', $email = true)
...
+			$commentEnd   = '\n //--> \n';
 		}
 
-		$replacement .= "\n //-->";
-		$replacement .= "\n </script>";
+		JFactory::getDocument()->addScriptDeclaration(
+		"
+		$commentStart
+		document.onreadystatechange = function () {
+			if (document.readyState == 'interactive') {
```
Comment (andrepereiradasilva) [inline libraries/cms/html/email.php]: will this work on all joomla supported browsers?

Comment (dgt41) [inline libraries/cms/html/email.php]: Should be according to Mozilla:  https://developer.mozilla.org/en-US/docs/Web/Events/readystatechange
(>= IE8)

**Comment by yvesh** [comment]
I have tested this item :white_check_mark: successfully on 351eb109a7d3235550c76fdb3567211b1230a243<br /><br />Works fine - thank you :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11027">issues.joomla.org/joomla-cms/11027</a>.</sub>

File: libraries/cms/html/email.php [hunk 2]
```diff
@@ -75,32 +67,56 @@ public static function cloak($mail, $mailto = true, $text = '', $email = true)
...
 		}
 
-		$replacement .= "\n //-->";
-		$replacement .= "\n </script>";
+		$doc->addScriptDeclaration(
+		"
+		$commentStart
+		document.onreadystatechange = function () {
+			if (document.readyState == 'interactive') {
+				document.getElementById('cloak$rand').innerHTML = '';
```
Comment (yvesh) [inline libraries/cms/html/email.php]: This inline $rand is really looking not nice :-)

File: libraries/cms/html/email.php [hunk 3]
```diff
@@ -75,32 +67,56 @@ public static function cloak($mail, $mailto = true, $text = '', $email = true)
...
 		{
-			$replacement .= "\n document.getElementById('cloak$rand').innerHTML += addy" . $rand . ";";
+			$tmpScript = "document.getElementById('cloak$rand').innerHTML += addy" . $rand . ";";
+		}
+
+		$doc          = JFactory::getDocument();
+		$commentStart = '';
+		$commentEnd   = '';
+
+		if ($doc->_type != 'html')
```
Comment (yvesh) [inline libraries/cms/html/email.php]: Any reason you are using _type here instead of getType()?

File: libraries/cms/html/email.php (hunk 2 above)
Comment (dgt41) [inline libraries/cms/html/email.php]: @yvesh you prefer:
```js
document.getElementById('cloak" . $rand . "').innerHTML = '';
```
this instead?

Comment (yvesh) [inline libraries/cms/html/email.php]: I would prefer that yes (Easier to set what's JS and what's PHP). What do you think?

Comment (dgt41) [inline libraries/cms/html/email.php]: Agree

File: libraries/cms/html/email.php (hunk 3 above)
Comment (dgt41) [inline libraries/cms/html/email.php]: Also agree on this one

**Comment by yvesh** [comment]
@dgt41 wouldn't that be a good moment to switch to something better then `random(1, 100000);` ?

There is a chance (even when it is low) that, when there are many emails on one page that you get the same id and run into troubles or leak the email.

Like with uniqid(rand()) or openssl random for example.

**Comment by dgt41** [comment]
@yvesh good idea, which one shall we use here?

**Comment by mbabker** [comment]
Don't explicitly couple it to openssl.  If you're worried about cryptographically secure randomness, use [random_int()](http://php.net/manual/en/function.random-int.php) (already available to us because we have the random_compat polyfill).  If you're not worried about it being cryptographically secure, either [rand()](http://php.net/manual/en/function.rand.php) or [mt_rand()](http://php.net/manual/en/function.mt-rand.php) should be more than enough.

The main thing it looks like you're trying to avoid is duplicating use of an ID, there's no way around that by changing the random number generator; you'd still need a static state somewhere saying "yes this number has already been used".

**Comment by yvesh** [comment]
`uniqid(rand())` should be more then unique enough (not even sure if rand() is neceassary, this is only needed when something is generated at the same microsecond).

As an alternative you could use `JUserHelper::genRandomPassword(10)`;

**Comment by dgt41** [comment]
How about: `// Random hash
		$rand = md5($mail . rand(1, 100000));`

**Comment by mbabker** [comment]
That'd definitely eliminate the need to track which numbers are in use.

**Comment by yvesh** [comment]
There is that one in a billion chance, that the same mail is printed multiple times on the page ;) But it should be fine. 

Hopefully md5 sums will be never so weak, that even email addresses are cracked in seconds. ;)

**Comment by mbabker** [comment]
You're not trying to cryptographically secure the addresses though.  It's really just making it more troublesome for bots to scrape them off the generated HTML page.  IMO it'd be better to use a different seed than the address you're trying to obfuscate, but in the long run all you're doing is inconveniencing some script kiddies crawling generated HTML.

**Comment by dgt41** [comment]
Reverse engineering open source code is a bit easier as you get to see the source code for free

**Comment by yvesh** [comment]
I have tested this item :white_check_mark: successfully on 7b2bb6e1beaeda0c728ca0261597f8bc968aa156<br /><br />Works fine and should be enough..<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11027">issues.joomla.org/joomla-cms/11027</a>.</sub>

File: libraries/cms/html/email.php [hunk 4]
```diff
@@ -75,32 +67,56 @@ public static function cloak($mail, $mailto = true, $text = '', $email = true)
...
 		{
-			$replacement .= "\n document.getElementById('cloak$rand').innerHTML += addy" . $rand . ";";
+			$tmpScript = "document.getElementById('cloak" . $rand . "').innerHTML += addy" . $rand . ";";
+		}
+
+		$doc          = JFactory::getDocument();
+		$commentStart = '';
+		$commentEnd   = '';
+
+		if ($doc->getType() != 'html')
```
Comment (andrepereiradasilva) [inline libraries/cms/html/email.php]: doesn't JDocument already does this?
https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/document/renderer/html/head.php#L170

Comment (dgt41) [inline libraries/cms/html/email.php]: @andrepereiradasilva you are right, I corrected that, thanks!

**Comment by yvesh** [comment]
I have tested this item :white_check_mark: successfully on 2df0786a556afcd45fdb7d662c3b309b825f53d7<br /><br />Still working<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11027">issues.joomla.org/joomla-cms/11027</a>.</sub>

**Comment by Devportobello** [comment]
Is this working in an ajax context ? (with injection in dom and js execution) ? unable to give it a try actually ...

**Comment by dgt41** [comment]
@Devportobello I guess we can still inject the script in the caller window with some magic to cover this case. can you point me to a module with an ajax call so I can provide the code needed?

**Comment by dgt41** [comment]
@Devportobello actually the script will not run if passed as data in the ajax call, so no B/C break here:
http://stackoverflow.com/questions/4619668/executing-script-inside-div-retrieved-by-ajax

**Comment by Devportobello** [comment]
@dgt41 you can execute JS with parseHTML : http://api.jquery.com/jQuery.parseHTML/
```javascript
// call ajax then in done function with response as arg:
$('<div>').append( $.parseHTML( response, document, true ) ).find( '.item-page' );
```

**Comment by dgt41** [comment]
@Devportobello can you share a url used in the ajax call (I have an idea) ?

**Comment by Devportobello** [comment]
@dgt41 sended by mail to you

**Comment by dgt41** [comment]
@Devportobello does my last commit cover the ajax part?

**Comment by Devportobello** [comment]
@dgt41 Nice try :), but in my case, i dont use "com_ajax" and i got 1 another custom tmpl "content", hmm diffcult to cover an ajax context, maybe with
```php
X-Requested-With : XMLHttpRequest
```

**Comment by Devportobello** [comment]
Sound good, will give a try when enough time

**Comment by killoltailored** [comment]
I have tested this item :white_check_mark: successfully on 863a9f28a0fd4552610f96783649b6a878a967ae<br /><br />I have tested it successfully 
After giving value "False" to javascript.enabled in FireFox 47.0.1 getting message "This email address is being protected from spambots. You need JavaScript enabled to view it."
Test in Joomla Version: Joomla! 3.6.0-rc Release Candidate [ Noether ] 28-June-2016 20:34 GMT 
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11027">issues.joomla.org/joomla-cms/11027</a>.</sub>

**Comment by AnishaTailored** [comment]
I have tested this item :white_check_mark: successfully on 863a9f28a0fd4552610f96783649b6a878a967ae<br /><br />Tested in Joomla! 3.6.0-rc2 [Noether] 6-July-2016 22:26 CET<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11027">issues.joomla.org/joomla-cms/11027</a>.</sub>

**Comment by hardiktailored** [comment]
Yes, the default functionality to show message works fine after applying patch but what about content updated by AJAX call? Give me example code to check. How can I check AJAX response disabling JavaScript?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11027">issues.joomla.org/joomla-cms/11027</a>.</sub>

**Comment by Devportobello** [comment]
Just tested rapidly, seem not working on ajax context...
Here is a module to test this PR: https://github.com/Devportobello/joomla-attach-PR-11027

**Comment by hardiktailored** [comment]
I have tested this item :red_circle: unsuccessfully on 863a9f28a0fd4552610f96783649b6a878a967ae<br /><br />Using AJAX to update content with email address in content seems not working. Shows message "This email address is being protected from spambots. You need JavaScript enabled to view it.", though JavaScript enabled.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11027">issues.joomla.org/joomla-cms/11027</a>.</sub>

File: libraries/cms/html/email.php [hunk 5]
```diff
@@ -75,32 +67,53 @@ public static function cloak($mail, $mailto = true, $text = '', $email = true)
...
 		{
-			$replacement .= "\n document.getElementById('cloak$rand').innerHTML += addy" . $rand . ";";
+			$tmpScript = "document.getElementById('cloak" . $rand . "').innerHTML += addy" . $rand . ";";
 		}
 
-		$replacement .= "\n //-->";
-		$replacement .= "\n </script>";
+		$inlineScript = '';
+		$script       = "
+		document.onreadystatechange = function () {
```
Comment (Devportobello) [inline libraries/cms/html/email.php]: not triggered in ajax context :/

**Comment by rahultomar** [comment]
I have tested this item :red_circle: unsuccessfully on 863a9f28a0fd4552610f96783649b6a878a967ae<br /><br />Same thing is happening after apply patch and before patch. Please more detail for testing.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11027">issues.joomla.org/joomla-cms/11027</a>.</sub>

**Comment by hardiktailored** [comment]
I have tested this item :white_check_mark: successfully on e95e92391a29f177f85e61e6b6b6a7a99ac086ff<br /><br />Context update by AJAX having email address in it, works now. I tested with module provided by @Devportobello <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11027">issues.joomla.org/joomla-cms/11027</a>.</sub>

**Comment by Devportobello** [comment]
Just noticed, there is no CS about inline JS in PHP ... http://joomla.github.io/coding-standards

**Comment by Devportobello** [comment]
I have tested this item :white_check_mark: successfully on e95e92391a29f177f85e61e6b6b6a7a99ac086ff<br /><br />I have tested this item ✅ successfully on e95e923

Work also in ajax context<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11027">issues.joomla.org/joomla-cms/11027</a>.</sub>

**Comment by brianteeman** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11027">issues.joomla.org/joomla-cms/11027</a>.</sub>

**Comment by dgt41** [comment]
@Devportobello I proposed to PLT to switch to this format
```php
JFactory::getDocument()->addScriptDeclaration(
<<<JS
//Script goes here with " and ' not needing escape
// Drawback you can't put vars like $this->id needs to be $id
JS
);
```
But is a lot of work to convert every inline script...

File: libraries/cms/html/email.php [hunk 6]
```diff
@@ -75,32 +67,58 @@ public static function cloak($mail, $mailto = true, $text = '', $email = true)
...
 		{
-			$replacement .= "\n document.getElementById('cloak$rand').innerHTML += addy" . $rand . ";";
+			$tmpScript = "document.getElementById('cloak" . $rand . "').innerHTML += addy" . $rand . ";";
 		}
 
-		$replacement .= "\n //-->";
-		$replacement .= "\n </script>";
+		$inlineScript = '';
+		$script       = "
+
```
Comment (roland-d) [inline libraries/cms/html/email.php]: @dgt41 Is there a reason for this empty line?

Comment (dgt41) [inline libraries/cms/html/email.php]: Nope!

**Comment by zero-24** [comment]
@joomla-cms-bot

**Pull request merged**